src/app/services/cross-table.service.ts

```diff
@@ -0,0 +1,442 @@
+  }
+
+  private retrieveAggregate(categoricalConceptConstraint: ConceptConstraint,
+                            peerConstraint: Constraint) {
+    if (categoricalConceptConstraint.concept.aggregate) {
+      const categoricalAggregate = <CategoricalAggregate>categoricalConceptConstraint.concept.aggregate;
+      this.composeCategoricalValueConstraints(categoricalAggregate, peerConstraint);
+    } else {
+      this.resourceService.getAggregate(categoricalConceptConstraint)
+        .subscribe((responseAggregate: Aggregate) => {
+          const categoricalAggregate = <CategoricalAggregate>responseAggregate;
+          this.composeCategoricalValueConstraints(categoricalAggregate, peerConstraint);
+        });
+    }
+  }
+
+  private composeCategoricalValueConstraints(categoricalAggregate: CategoricalAggregate,
+                                             peerConstraint: Constraint) {
+    let categories = categoricalAggregate.values;
+    for (let category of categories) {
+      let val = new ValueConstraint();
+      val.valueType = 'STRING';
+      val.operator = '=';
+      val.value = (category === FormatHelper.nullValuePlaceholder) ? null : category;
+      val.textRepresentation = val.value.toString();
+      let combi = new CombinationConstraint();
+      combi.addChild(peerConstraint);
+      combi.addChild(val);
+      combi.textRepresentation = this.adjustCombinationConstraintTextRepresentation(combi);
+      this.crossTable.addHeaderConstraint(peerConstraint, combi);
+    }
+    this.updateRows();
+  }
+
+  private adjustCombinationConstraintTextRepresentation(constraint: CombinationConstraint): string {
+    let description = constraint.textRepresentation;
+    if (constraint.isAnd()) {
+      let numValueConstraints = 0;
+      let numCatConceptConstraints = 0;
+      let valChild = null;
+      let catChild = null;
+      constraint.children.forEach((child: Constraint) => {
+        if (child.getClassName() === 'ValueConstraint') {
+          numValueConstraints++;
+          valChild = child;
+        } else if (this.constraintService.isCategoricalConceptConstraint(child)) {
+          numCatConceptConstraints++;
+          catChild = child;
+        }
+      });
+      if (numValueConstraints === 1) {
+        description = valChild.textRepresentation;
+      } else if (numCatConceptConstraints === 1) {
+        description = catChild.textRepresentation;
+      }
+    }
+
+    return description;
+  }
+
+  get areHeaderConstraintsMapped(): boolean {
+    let mapped = true;
+    this.rowConstraints.forEach((con: Constraint) => {
+      let hasIt = this.headerConstraints.has(con);
+      if (!hasIt) {
+        mapped = hasIt;
+      }
+    });
+    if (mapped) {
+      this.columnConstraints.forEach((con: Constraint) => {
+        let hasIt = this.headerConstraints.has(con);
+        if (!hasIt) {
+          mapped = hasIt;
+        }
+      });
+    }
+    return mapped;
+  }
+
+  mockDataInit() {
```

Comment:
this part is now duplicated in cross-table.service.mock.ts

Reply:
yes, the two are rather similar, but one is using dimensions the other is using constraints